Convert [ParticleProto] to [Plan.Particle].

bgogul opened the pull request:
One more to go convert recipe.proto to the Kotlin classes!

jasonwyatt commented on java/arcs/core/data/proto/ParticleProtoDecoder.kt:
```diff
@@ -0,0 +1,57 @@
+    val particleSpec = context.particleSpecs[specName]
+    require(particleSpec != null) {
+        "ParticleSpec '$specName' not found in decoding context."
+    }
```
```suggestion
    val particleSpec = requireNotNull(context.particleSpecs[specName]) {
        "ParticleSpec '$specName' not found in decoding context."
    }
```

jasonwyatt commented on java/arcs/core/data/proto/ParticleProtoDecoder.kt:
```diff
@@ -0,0 +1,57 @@
+    val handleProto = context.handleProtos[handle]
+    require(handleProto != null) {
+        "HandleProto for '$handle' not found when decoding ParticleProto '${particleSpec.name}'."
+    }
```
```suggestion
    val handleSpec = requireNotNull(particleSpec.connections[name]) {
        "HandleConnectionSpec '$name' not found in ParticleSpec '${particleSpec.name}'."
    }
    val handleProto = requireNotNull(context.handleProtos[handle]) {
        "HandleProto for '$handle' not found when decoding ParticleProto '${particleSpec.name}'."
    }
```

jasonwyatt commented on java/arcs/core/data/proto/ParticleProtoDecoder.kt:
```diff
@@ -0,0 +1,57 @@
+    }
+    val handleConnections = connectionsList.map {
+        it.handle to it.decode(particleSpec, context)
+    }.toMap()
```
Consider using `associateBy` or one of the related methods.

bgogul replied:
Unfortunately, the decoded `HandleConnection` does not have the `name` that I can use to `associateBy`. I have marked adding `name` as one of the TODOs.

jasonwyatt commented on javatests/arcs/core/data/proto/ParticleProtoDecoderTest.kt:
```diff
@@ -0,0 +1,126 @@
+    val context = DecodingContext(
+        particleSpecs = mapOf("Reader" to readerSpec, "Writer" to writerSpec),
+        handleProtos = mapOf("thing" to thingHandleProto)
+    )
```
These can go at the top of the file (they're not in a companion object, so they're just instance fields)

bgogul commented:
> Some readability suggestions, but nothing huge :)

Thanks, Jason. Addressed your comments. I added some more tests involving multiple connections. PTAL.

bgogul commented:
Thanks for the review, folks! However, I need to abandon this PR. The `Plan.Particle` does not allow me the capture information that is  necessary for type inference and other analysis. recipe.proto has all the inforamation. Therefore, I will introduce a `Recipe` type in Kotlin for the purposes of the analysis.  Once types are resolved, the recipe can be converted to a `Plan`.
